### PR TITLE
Allow C code to interface, at a basic level, to code in ykrt.

### PR DESCRIPTION
**Pull request by ltratt**
This is crude, and will need substantial expanding (and probably tweaking), but it's a start.

### CODE REVIEW

File: c_tests/run.rs [hunk 1]
```diff
@@ -12,22 +12,29 @@ const COMMENT: &str = "//";
 /// Make a compiler command that compiles `src` to `exe`.
 fn mk_compiler(exe: &Path, src: &Path) -> Command {
     let mut compiler = Command::new("clang");
+
```
Comment (vext01) [inline c_tests/run.rs:15]: Some unrelated whitespace shuffling here.

File: ykrt/ykrt.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+// Program positions that can't be control points don't need an associated
+// `Location`. For interpreters that can't (or don't want) to be as selective,
+// a simple (if moderately wasteful) mechanism is for every bytecode or AST
+// node to have its own `Location` (even for bytecodes or nodes that can't be
+// control points).
+typedef struct {
+    uintptr_t state;
+} Location;
+
+#if defined(__x86_64)
```
Comment (vext01) [inline ykrt/ykrt.h:24]: Would it be better to do the test based on pointer size rather than platform?

File: ykrt/ykrt.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+// Return a reference to the global `MT` instance: at any point, there is at
+// most one of these per process and an instance will be created if it does not
+// already exist.
+MT *yk_mt();
+
+// Return this `MT` instance's current hot threshold. Notice that this value
+// can be changed by other threads and is thus potentially stale as soon as it
+// is read.
+HotThreshold yk_mt_hot_threshold(MT *);
```
Comment (vext01) [inline ykrt/ykrt.h:38]: Is it important to expose the current hot threshold?

File: ykrt/ykrt.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+//
+// Note that a `Location` created by this call must not simply be discarded: if
+// no longer wanted, it must be passed to `yk_drop_location` to allow
+// appropriate clean-up.
+Location yk_new_location();
+
+// Clean-up a `Location` previously created by `yk_new_location`. The
+// `Location` must not be further used after this call or undefined behaviour
+// will occur.
+void yk_drop_location(Location *);
```
Comment (vext01) [inline ykrt/ykrt.h:55]: The convention we've been using to free things from C is `free` (not `drop`). We should probably be consistent.

**Comment by vext01** [comment]
Mostly LGTM. Some of the changes are unrelated and they'd have been better suited to separate commits. I won't make you unpick it all though :P

File: c_tests/run.rs (hunk 1 above)
Comment (ltratt) [reply c_tests/run.rs:15]: Agreed.

File: ykrt/ykrt.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+// `Location` does not need to be at a stable address in memory and can be
+// freely moved.
+//
+// Program positions that can't be control points don't need an associated
+// `Location`. For interpreters that can't (or don't want) to be as selective,
+// a simple (if moderately wasteful) mechanism is for every bytecode or AST
+// node to have its own `Location` (even for bytecodes or nodes that can't be
+// control points).
+typedef struct {
+    uintptr_t state;
```
[the remark this answers is not in the document]
Comment (ltratt) [reply ykrt/ykrt.h:21]: Rust, unfortunately, conflates "the size of an int" with "the size of a pointer". Modern C differentiates them as `size_t` and `uintptr_t`: since `Location`s encode a pointer, the latter is the right thing for us.

File: ykrt/ykrt.h (hunk 2 above)
Comment (ltratt) [reply ykrt/ykrt.h:24]: I couldn't work out how to do that portably :/ This was my hack on the basis that it fails fast: if `__X86_64` isn't defined you get a compile-time error.

File: ykrt/ykrt.h (hunk 3 above)
Comment (ltratt) [reply ykrt/ykrt.h:38]: Yes and no: it's not *vital* but without it you can't write tests which exceed the hot threshold reliably, so it is fairly useful.

File: ykrt/ykrt.h (hunk 4 above)
Comment (ltratt) [reply ykrt/ykrt.h:55]: This isn't `free`ing the memory being pointed to though: `free` and `drop` are subtly different IMHO.

Comment (vext01) [reply ykrt/ykrt.h:55]: I see it as the same thing just in different languages terminology.

I'm OK with drop if we go and change all of the other instances in the repo.

File: ykrt/ykrt.h (hunk 3 above)
Comment (vext01) [reply ykrt/ykrt.h:38]: ok

File: ykrt/ykrt.h (hunk 4 above)
Comment (vext01) [reply ykrt/ykrt.h:55]: Lets move on and come back to it.

File: ykrt/ykrt.h (hunk 2 above)
Comment (vext01) [reply ykrt/ykrt.h:24]: You're right: https://scaryreasoner.wordpress.com/2009/02/28/checking-sizeof-at-compile-time/

I guess we leave it as-is then.

File: ykrt/ykrt.h (hunk 4 above)
Comment (ltratt) [reply ykrt/ykrt.h:55]: Agreed.

File: ykrt/ykrt.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (ltratt) [reply ykrt/ykrt.h:21]: > I thought that Location may encode a pointer.

Yes... but I'm not sure what deeper question you're asking?

[the remark this answers is not in the document]
Comment (vext01) [reply ykrt/ykrt.h:21]: Yeah, I was confused.
